Workaround for Dataframe object in the config: check whether value is string before mandatory missing check

jeffknaide opened the pull request:
Closes #566 

sidestep source of https://github.com/facebookresearch/hydra/issues/1285 at the omegaconf level

Jasha10 commented on omegaconf/basecontainer.py:
```diff
@@ -784,7 +784,7 @@ def _create_structured_with_missing_fields(
     if isinstance(value, Container):
         value = value._value()
-    ret = value == "???"
+    ret = value == "???" if isinstance(value, str) else False
```
So the idea is to sidestep the custom `DataFrame.__eq__` method by first deciding if `value` is a `str`, then only calling `value.__eq__("???")` if we are sure that `value` is a `str`.

jeffknaide replied:
yep, exactly.

omry commented:
Thanks @jeffknaide.
We have some outstanding discussions around the internal behavior of MISSING, we will address this once we comes to a conclusion there.

omry commented:
Blocked by #543.

omry commented on omegaconf/_utils.py:
```diff
@@ -353,6 +353,9 @@ def get_value_kind(
 
     value = _get_value(value)
 
+    if not isinstance(value, str):
+        return ValueKind.VALUE
+
```
Change the following check to:
```
    if isinstance(value,str) and value == "???":
        return ValueKind.MANDATORY_MISSING
```
instead.

omry commented on omegaconf/basecontainer.py:
```diff
@@ -784,7 +784,7 @@ def _create_structured_with_missing_fields(
     if isinstance(value, Container):
         value = value._value()
-    ret = value == "???"
+    ret = value == "???" if isinstance(value, str) else False
```
```suggestion
    ret = isinstance(value, str) and value == "???"
```

omry commented on tests/__init__.py:
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
+    def __init__(self) -> None:
+        pass
+
+    def __eq__(self, other: Any) -> Any:
+        return self
```
I find `__eq__` returning self very odd. what's going on here?

jeffknaide replied:
when a pandas DataFrame runs `df.__eq__(val)`, it returns a DataFrame of booleans with the same shape as the original DataFrame. Here I was trying to mimic the behavior of a df comparison resulting in the same class being returned, as this is partially what caused issues originally. perhaps this should be made clearer with a comment?

omry replied:
Gotcha.
Yes, add a comment what the behavior we are mimicking is.

jeffknaide replied:
i don't think so -- if we don't override eq, the comparison to string would return a bool and then not raise the expected error on the conditional. eg `if Dataframe() == "???"` would equate to `if False` (which wouldn't raise the expected error) without the eq override.

omry commented on tests/__init__.py:
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
         raise NotImplementedError()
 
 
+class Dataframe:
```
Add a comment here.
something like:

```
Emulates Pandas Dataframe equality behavior.
```

omry commented on tests/__init__.py:
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
+        return self
+
+    def __bool__(self) -> None:
+        raise ValueError
```
add a comment explaining what this achieves.

jeffknaide replied:
let me know if the comments I added here work

omry commented on tests/test_merge.py:
```diff
@@ -599,6 +600,14 @@ def test_merge_allow_objects(merge: Any) -> None:
     assert ret == {"a": 10, "foo": iv}
 
 
+def test_merge_with_allow_Dataframe() -> None:
+    kwargs = OmegaConf.create({"a": Dataframe()}, flags={"allow_objects": True})
+    cfg = OmegaConf.create()
+
+    cfg.merge_with(kwargs)
+    assert isinstance(cfg.a, Dataframe)
```
If this still reproduces the problem, please switch to it:
```suggestion
def test_merge_with_allow_Dataframe() -> None:
    cfg = OmegaConf.create({"a": Dataframe()}, flags={"allow_objects": True})
    ret = OmegaConf.merge({}, cfg)
    assert isinstance(ret.a, Dataframe)
```

omry commented on tests/test_nodes.py:
```diff
@@ -559,3 +559,10 @@ def test_allow_objects() -> None:
     iv = IllegalType()
     c.foo = iv
     assert c.foo == iv
+
+
+def test_allow_Dataframe_objects() -> None:
+    c = OmegaConf.create()
+    c._set_flag("allow_objects", True)
+    c.a = Dataframe()
+    assert isinstance(c.a, Dataframe)
```
allow_objects is already tested, remove this test.

omry reviewed (changes requested):
Thanks @jeffknaide.
See inline comments.

omry commented:
lgtm.
@odelalleau, this could interact with your outstanding PR to change the semantics of missing.

odelalleau commented:
> lgtm.
> @odelalleau, this could interact with your outstanding PR to change the semantics of missing.

At first glance I don't foresee any issue (*crossing fingers*). Just mentioning that maybe there could be a new utility function to avoid repeating `if isinstance(x, str) and x == "???"` everywhere?

omry commented:
Yeah, was thinking that too.

jeffknaide commented:
@odelalleau @omry oh, yes, that makes sense. I just added `is_mandatory_missing` to `_utils` and referenced it where i had been using `if isinstance(x, str) and x == "???"`.

There was also an inner function defined in `BaseContainer._resolve_with_default` called `is_mandatory_missing` which was returned of `get_value_kind`. I just removed that inner function in favor of `is_mandatory_missing` defined in _utils. This seems to break the tests, though, so I must be missing something that's happening within `get_value_kind`

odelalleau commented:
> I must be missing something that's happening within `get_value_kind`

It’s almost certainly because that function can be called on a node: it starts with `_get_value(..)` to obtain the node’s value if that’s the case. So in BaseContainer the problem is likely related to ˋ if is_mandatory_missing(resolved):`, and replacing ˋresolved` with its value here should fix it.

omry commented:
I think we will want to take a closer look at this in a followup.
can you undo the refactoring changes?

odelalleau commented:
> can you undo the refactoring changes?

Most could be kept, only the change in `_resolve_with_default()` needs reverting

omry commented:
Maybe. it just feel like there are a bit too many layers here. The logic can be slightly reorganized to be simpler.

jeffknaide commented:
@omry - i just reverted previous commit. let me know if you want me to put the utils function back while leaving `_resolve_with_default ` alone.

omry reviewed (approved):
Thanks @jeffknaide.
